test/lit/passes/global-subtyping.wast

```diff
@@ -0,0 +1,538 @@
+  ;; CHECK-NEXT:  (struct.set $struct 0
+  ;; CHECK-NEXT:   (local.get $struct)
+  ;; CHECK-NEXT:   (local.get $child-A)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $work (param $struct (ref $struct)) (param $child-A (ref $child-A))
+    (struct.set $struct 0
+      (local.get $struct)
+      (local.get $child-A)
+    )
+    (struct.set $struct 0
+      (local.get $struct)
+      (local.get $child-A)
+    )
+  )
+)
+
+(module
+  ;; Write to the parent a child, and to the child a parent. The write to the
+  ;; child prevents specialization even in the parent and we only improve up to
+  ;; $struct but not to $child.
+
+  ;; CHECK:      (type $struct (struct_subtype (field (mut (ref $struct))) data))
+  (type $struct (struct_subtype (field (mut dataref)) data))
+
+  ;; CHECK:      (type $child (struct_subtype (field (mut (ref $struct))) $struct))
+  (type $child (struct_subtype (field (mut dataref)) $struct))
+
+  ;; CHECK:      (type $ref|$struct|_ref|$child|_=>_none (func_subtype (param (ref $struct) (ref $child)) func))
+
+  ;; CHECK:      (func $work (type $ref|$struct|_ref|$child|_=>_none) (param $struct (ref $struct)) (param $child (ref $child))
+  ;; CHECK-NEXT:  (struct.set $struct 0
+  ;; CHECK-NEXT:   (local.get $struct)
+  ;; CHECK-NEXT:   (local.get $child)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.set $child 0
+  ;; CHECK-NEXT:   (local.get $child)
+  ;; CHECK-NEXT:   (local.get $struct)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $work (param $struct (ref $struct)) (param $child (ref $child))
+    (struct.set $struct 0
+      (local.get $struct)
+      (local.get $child)
+    )
+    (struct.set $child 0
+      (local.get $child)
+      (local.get $struct)
+    )
+  )
+)
+
+(module
+  ;; As above, but both writes are of $child, so we can optimize.
+
+  ;; CHECK:      (type $child (struct_subtype (field (mut (ref $child))) $struct))
+  (type $child (struct_subtype (field (mut dataref)) $struct))
+
+  ;; CHECK:      (type $struct (struct_subtype (field (mut (ref $child))) data))
+  (type $struct (struct_subtype (field (mut dataref)) data))
+
+  ;; CHECK:      (type $ref|$struct|_ref|$child|_=>_none (func_subtype (param (ref $struct) (ref $child)) func))
+
+  ;; CHECK:      (func $work (type $ref|$struct|_ref|$child|_=>_none) (param $struct (ref $struct)) (param $child (ref $child))
+  ;; CHECK-NEXT:  (struct.set $struct 0
+  ;; CHECK-NEXT:   (local.get $struct)
+  ;; CHECK-NEXT:   (local.get $child)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.set $child 0
+  ;; CHECK-NEXT:   (local.get $child)
+  ;; CHECK-NEXT:   (local.get $child)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $work (param $struct (ref $struct)) (param $child (ref $child))
+    (struct.set $struct 0
+      (local.get $struct)
+      (local.get $child)
+    )
+    (struct.set $child 0
+      (local.get $child)
+      (local.get $child)
+    )
+  )
+)
+
+(module
+  ;; As in GP, write to the parent a child, and to the child a parent, but now
```

Comment:
What is `GP`?

Reply:
Clearly I spend too much time online... "GrandParent" (or really "Grandparent Post[er]"). That is, not the preceding testcase but one before it. I'll clarify.

Reply:
I've only ever seen "OP" in the wild. Clearly I need to spend more time online!